---
title: Metagenomics Software Setup Protocol
author: Guangyu Li
date: Aug 2017
output: html_document
---


# Overview {#top}

This protocol helps setting up a basic metagenomics computational environment on NEU's Discovery cluster (Linux, and with no administrative (sudo) privileges). Note anyone tries to install those software manually as shown in this protocol is assumed to have basic knowledge of Linux, mandatory skill set of working with terminal (bash) and know how to connect to a remote machine and run commands via ssh. Before the installation, please read the [general guidance](#general) first. Installing programs are in this [list](#install).


# General guidance about working on Discovery cluster {#general}

### Managing space

Your home directory (/home/you_account_name/) on Discovery cluster (which is your "own space") has a 30GB limit. It is good to install your softwares into, but not run any jobs or build any databases in the home directory. Instead, you should create your own scratch directory in /scratch/your_account_name/. The scratch partition is 1.1PB (1PB=1024TB, 1TB=1024GB) in size and should be sufficient to run most of the jobs. But still make sure that clean the temporary files in your scratch directory after each job is done.

### Running jobs

Once you connected to the Discovery cluster (via discover2.neu.edu or discovery4.neu.edu), you are on "login nodes". These nodes are not designed of running cpu-intensive jobs, so <b>DO NOT RUN ANY JOBS ON THESE NODES</b>. To run jobs, you should allocate an "interactive node" or submit as a "batch job" to the computing nodes (discribed in detail [here](https://www.northeastern.edu/rc/?page_id=18)). However, most computing nodes don't have the access to internet. You have to <b>downloading data on login nodes</b>, or download locally then scp to the remote machine.

### Organizing your software

Everyone has his own manner of organizing files. On Linux systems, a common way of installing software locally is creating a directory called "local" in your home directory, and assign it as prefix whenever installing a software. This protocol is written assuming this manner of organizing, which is better for general use; however, if you copy my build directly, the organizing is different, which, in result will be a bit easier when multiple versions coexist. 

<font color="#00FF00"><b>Guide:</b></font> If you choose the first way, running these codes first before installing any software:

```
$ if [[ !-d ~/local ]]; then mkdir ~/local; fi
$ if [[ !-d ~/local/bin ]]; then mkdir ~/local/bin; fi
$ if [[ !-d ~/local/include ]]; then mkdir ~/local/include; fi
$ if [[ !-d ~/local/lib ]]; then mkdir ~/local/lib; fi
$ if [[ !-d ~/Programs ]]; then mkdir ~/Programs; fi # this directory holds all other scripts come with all packages
```

<font color="#00FF00"><b>Note:</b></font> Feel free to explore [NEU's research computing website](https://www.northeastern.edu/rc/) for more information, policies and tutorials about working with Discovery cluster.

[return to top](#top)


# List of installing software {#install}


### What's already on Discovery cluster

These software are already installed and ready to use in general. If multiple versions of a software are present, the newest version is listed below. Some of the software may be a bit out of date, but still operational in most cases.

* GCC 5.2.1
* GNU C/C++ library (glibc/glibcxx)
* GNU Scientific Library (GSL)
* SQLite3
* Perl 5.20.0
* R 3.3.3
* Python 2.7.5
* Python 3.5.2
* Oracle JDK 1.8
* SAMtools 0.1.19
* Bowtie 2.1.0
* BLAST+ 2.2.28
* Libraries (HDF5, etc.)
* Perl modules (BioPerl)
* Python2 libs (pip, NumPy, Scipy, matplotlib, Sklearn, BioPython, etc.)
* Python3 libs (pip, NumPy, Scipy, matplotlib, Sklearn, BioPython, etc.)

<font color="#FF0000"><b>Important:</b></font> The Discovery cluster is found pool in organizing coexisting Python2 and Python3 (Python3 is not the superset of Python2!). Loading these two versions simultaneously causes path problem (specifically by Python2 module). So when run some scripts that require loaded modules, load single module at a time and reset your bash environment before doing next steps! Another solution requires more Linux knowledge and efforts, to install a virtual envrionment and install your own modules inside. However, this too will cause some issues (e.g., matplotlib).

### Utility software

<font color="#0000FF"><b>Optional:</b></font> These software are optional. For general use, I recommend <b>NOT</b> install them unless specific requirements needed (i.e. in case very new software is needed which fails compilation using existing gcc (5.2.1); such case, however, is very rare. The gcc 5.2.1 should satisfy most time.)

* [GCC 5.4.0 (optional)](#gcc)
* [GNU binutils (optional)](#binutils)

### Binary programs and dependencies {#load_gcc}

<font color="#FF0000"><b>Important:</b></font> Due to the poor maintenance of Discovery cluster, loaded modules will cause problems in finding programs. Hence make sure that no other modules except the three mandatory ones (perl, fftw and slurm) are loaded, then in terminal, input:

```
$ module load gnu-5.2.1-OpenACC_2.0-Cuda7.0
```

to load the gcc 5.2.1 compiler before installing software in this list. If you already build the gcc 5.4.0 in the above list, just skip this.

* [Aragorn](#aragorn)
* [Barrnap](#barrnap)
* [Bcftools](#bfctools)
* [Bedtools](#bedtools)
* [BLAST+](#blast)
<!-- * [bowtie](#bowtie) -->
* [BWA (Burrows-Wheeler Aligner)](#bwa)
* [Bzip2](#bzip2)
* [Diamond](#diamond)
* [Kraken](#kraken)
* [Hmmer](#hmmer)
* [HTSlib](#htslib)
* [Infernal](#infernal)
* [Jellyfish](#jellyfish)
* [GNU Parallel](#parallel)
* [Prodigal](#prodigal)
* [Prokka](#prokka)
* [SAMtools](#samtools)
* [tbl2asn](#tbl2asn)
* [XZ Utils](#xzutils)
* [Zlib](#zlib)

### Java packages

* [Picard (Java)](#picard)

### Python packages

* [Anvio (Python3)](#anvio)
* [CONCOCT (Python2)](#concoct)
<!-- * [MetaPhlAn](#metaphlan) -->
* [Pysam](#pysam)
* [Quast](#metaphlan)

### Perl packages

* [List::MoreUtils](#list_moreutils)
* [XML::Simple](#xml_simple)

<font color="#00FF00"><b>Note:</b></font> Some software comes only with source code. The users have to compile them on their own. This is the reason why setting up this environment is so painful. If you are impatient, you can follow [this instruction](#impatient) to directly copy my build to your own, skipping the annoing compilation.

<font color="#00FF00"><b>Note:</b></font> During compiling, warnings can be ignored safely in almost all cases.


# Installation guide


## For impatient people {#impatient}

If you only want operational software (hopefully they will work), you can just copy my build and path settings directly by running (unfortunately you still need to be patient because it takes a long long while):

```
$ cp -r /home/li.gua/usr/ ~/
$ cp -r /home/li.gua/Jar/ ~/
$ cp -r /home/li.gua/Programs/ ~/
$ cp /home/li.gua/.ncbirc ~/
$ cat /home/li.gua/.bash_paths >> ~/.bashrc
$ . ~/.bashrc
```

[return to top](#top)

## Full installation guide

### > Aragorn {#aragorn}

* <font color="#808080">Supports:</font> [Prokka](#prokka).
* <font color="#00FF80">Simple:</font> Simple compilation and installation.

Download aragorn 1.2.38 [here](http://mbio-serv2.mbioekol.lu.se/ARAGORN/Downloads/). Make sure you install version >= 1.2 anyway.

For installation:

```
$ tar -xvzf aragorn1.2.38.tgz
$ cd aragorn1.2.38
$ cd gcc -O3 -ffast-math -finline-functions -o aragorn aragorn1.2.38.c
$ cp aragorn ~/local/bin/
```

[return to top](#top)

### > Barrnap {#barrnap}

* <font color="#808080">Supports:</font> [Prokka](#prokka).
* <font color="#00FF80">No-compile:</font> No need to compile.
* <font color="#0000FF">Perl:</font> An executable Perl script.

Visit [project home](http://www.vicbioinformatics.com/software.barrnap.shtml) or [GitHub repository](https://github.com/tseemann/barrnap). The latest version is 0.8.

If you download from the GitHub release (v0.8), it is already compiled. Just need to copy the executable:

```
$ unzip barrnap-0.8.zip
$ cd barrnap-0.8
$ cp ./bin/barrnap ~/local/bin/
$ cp -r ./db ~/local/ # copy this db to this location; however, you can copy to anywhere else and change its path by opening barrnap in vim (it's basically a perl program)
$ #cp ./binaries/linux/nhmmer ~/local/bin/ # this is necessary in some cases; however we will install hmmer anyway, so this is not necessary (it comes with hmmer, too)
```

[return to top](#top)

### > Bcftools {#bcftools}

* <font color="#00FFFF">Dependencies:</font> [HTSlib](#htslib).

[Bcftools](#bcftools), [HTSlib](#htslib) and [SAMtools](#samtools) come as a suite. If you try to install more than one of them, it is better to install the same version across. The repository page on GitHub is [here](https://github.com/SAMtools). The latest version so far is 1.5, on release page [here](https://github.com/SAMtools/bcftools/releases). SAMtools are already built on Discovery cluster, however, it is relatively low in version (0.1.19).

For installation:

```
$ tar -jxvf bcftools-1.5.tar.bz2
$ cd bcftools-1.5
$ ./configure --prefix="$HOME/local" CFLAGS="-I$HOME/local/include" LDFLAGS="-L$HOME/local/lib" --with-htslib="$HOME/local"
$ make
$ make install
```

### > Bedtools {#bedtools}

* <font color="#FF0000">Attention:</font> Compiling with special attention.
* <font color="#00FFFF">Dependencies:</font> [Zlib](#zlib).

Visit [project home](http://bedtools.readthedocs.io/en/latest/index.html) or download source code at [GitHub repository](https://github.com/arq5x/bedtools2/). The latest release is v2.26.0.

For installation:

```
$ unzip bedtools2-2.26.0.zip
$ cd bedtools2-2.26.0
```

The Bedtools speficically requires the static lib of zlib (libz.a) which cannot be found on Discovery cluster. To continue, make sure you build your own [Zlib](#zlib); You can also just use my build with replacing the path (things after -L) by /home/li.gua/usr/zlib/1.2.11/lib. To move on:

```
make LDFLAGS="-L$HOME/local/lib"
make install --prefix="$HOME/local"
```

<!-- <font color="#FF0000">Attention:</font> There is a bug in makefile rules, which must be fixed before going further. Using vim to edit the Makefile:

```
$ vim Makefile
```

Find this line (Line 102) using arrow keys:

```
BUILT_OBJECTS = $(OBJ_DIR)/*.o
```

<font color="#FF0000"><b>Important:</b></font> Please follow exactly the instruction, or you will easily get lost if have no experience on Vim.

(1) press <b>i</b> on keyboard (thus your are in editing mode, feeling the same as working with ordinary text editor)
(2) change this line into (specify the the libz.a library; <font color="#00FF00"><b>Guide:</b></font> if you prefer using my build, just add /home/li.gua/usr/zlib/1.2.11/lib/libz.a):

```
BUILT_OBJECTS = $(OBJ_DIR)/*.o /home/your_account_name/local/lib/libz.a
```

(3) After done the chage, press <b>ESC</b>, <b>:wq</b> then press <b>ENTER</b> to save the file and quit vim.

Then it is free to continue building and installing bedtools by running:

```
make
make install --prefix="$HOME/local"
``` -->

[return to top](#top)

### > BLAST+ {#blast}

* <font color="#00FF80">No-compile:</font> No need to compile.
* <font color="#00FFFF">Dependencies:</font> [List::MoreUtils](#list_moreutils).
* <font color="#FF8000">Special:</font> Need to build databases after successful installation.

Download the latest version (2.6.0) [here](ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.6.0+-x64-linux.tar.gz) via [NCBI](https://blast.ncbi.nlm.nih.gov/Blast.cgi).

For installation (simply decompress the archive and copy):

```
$ tar -zxvf ncbi-blast-2.6.0+-x64-linux.tar.gz
$ cp -r ncbi-blast-2.6.0+/ ~/local/blast/
```

Don't forget to add it to your PATH environment variable, if you don't know how to do it, check below:

```
$ echo 'export PATH="$HOME/local/blast/bin:$PATH"' >> ~/.bashrc
```

<font color="#00FF00"><b>Note:</b></font> If you want to do local blast, the database is needed, otherwise you do not need to do so (by running against NCBI's remote database with switch -remote if you have internet access).

<font color="#00FF00"><b>Note:</b></font> This database is huge. To save space on Discovery cluster, you can check and use my database at /scratch/li.gua/BLAST_DB/ if available. By adding below information to your ~/.ncbirc (create this file if not exist):

```
[BLAST]
BLASTDB=/scratch/li.gua/BLAST_DB
```

If you want to build the database anyway, try this:

```
cd /path/to/your/database/
update_blastdb.pl nr swissprot pdbaa refseq_protein taxdb nt

# this step can be done separately after download, on any interactive nodes:
tar -zxpf *.tar.gz
```

These databases should be roughly enough. You may still need to edit your ~/.ncbirc file (see above step).

Other libraries that cannot be built with update_blastdb.pl:

* [Cdd](ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Cdd_LE.tar.gz) (NCBI's Conserved Domains Database)
* [COG](ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Cog_LE.tar.gz) (NCBI's Clusters of Orthologous Groups Database)
* [Pfam](ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Pfam_LE.tar.gz)
* [Smart](ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/little_endian/Smart_LE.tar.gz)

All these are in little endians; for big endians, check this [FTP](ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/big_endian/). Download all tarballs into your $BLASTDB and proceed with untar. These databases are all comparable with other NCBI blast databases.

[return to top](#top)

### > BWA (Burrows-Wheeler Aligner) {#bwa}

Visit [project home](http://bio-bwa.sourceforge.net/) or download source code [here](https://sourceforge.net/projects/bio-bwa/files/bwa-0.7.16a.tar.bz2/download). If you are looking at the project page, make sure <b>DO NOT</b> download the bwakit-0.7.12_x64-linux.tar.bz2, it's actually out of date.

For installation:

```
$ tar -jxvf bwa-0.7.16a.tar.bz2
$ cd bwa-0.7.16a
$ make
$ cp bwa ~/local/bin/
$ cp -r bwa/bwakit/ ~/local/bin/
```

[return to top](#top)

### > Bzip2 {#bzip2}

* <font color="#808080">Supports:</font> [bowtie](#bowtie), [HTSlib](#htslib).
* <font color="#FF0000">Attention:</font> Compiling with special attention.

Its project home is [here](http://www.bzip.org/), and latest source tarball should be found in [this link](http://www.bzip.org/1.0.6/bzip2-1.0.6.tar.gz).

For installation:

```
$ tar -zxvf bzip2-1.0.6.tar.gz
$ cd bzip2-1.0.6
$ make -f Makefile-libbz2_so # we need a shared library. directly compile with make will end up in a static object
$ make install PREFIX="$HOME/local"
```

[return to top](#top)

### > Diamond {#diamond}

* <font color="#00FF80">No-compile:</font> No need to compile.

Download precompiled binaries with [this link](https://github.com/bbuchfink/diamond/releases/download/v0.9.10/diamond-linux64.tar.gz) or visit repository page on [GitHub](https://github.com/bbuchfink/diamond).

For installation (simply untar the compressed archive and copy):

```
$ tar -zxvf diamond-linux64.tar.gz
$ cp diamond ~/local/bin/
```

This release also includes a manual (PDF) file.

[return to top](#top)

### > Kraken {#kraken}

* <font color="#00FF80">Simple:</font> Simple compilation and installation.
* <font color="#00FFFF">Dependencies:</font> [Jellyfish](#jellyfish).
* <font color="#FF8000">Special:</font> Need to build databases after successful installation.

Visit the project home and download the source code [here](https://ccb.jhu.edu/software/kraken/). Note compile and install don't require [Jellyfish](#jellyfish), but building the library does.

```
$ tar -zxvf kraken-0.10.5-beta.tgz
$ cd kraken-0.10.5-beta
$ ./install_kraken.sh ~/local/kraken
$ ln -s -T "$HOME/local/kraken/kraken" "$HOME/local/bin/kraken"
$ ln -s -T "$HOME/local/kraken/kraken-build" "$HOME/local/bin/kraken-build"
```

For build the library (bacteria, viruses and human with taxonomy) into scratch directory (beware don't accidentally delete it!). First download the data from NCBI using these codes (this will take a while):

```
$ cd /scratch/your_account_name/
$ mkdir KRAKEN_DB && mkdir KRAKEN_DB/STANDARD
$ cd KRAKEN_DB
$ kraken-build --db STANDARD --download-taxonomy
$ kraken-build --db STANDARD --download-library bacteria
$ #kraken-build --db STANDARD --download-library plasmids # Currently it is not available via NCBI. You can still try this out though.
$ kraken-build --db STANDARD --download-library viruses
$ kraken-build --db STANDARD --download-library human
```

Then allocate an interactive node (adjust your thread setting accordingly) and run codes below (this will take more while):

```
$ cd /scratch/your_account_name/KRAKEN_DB
$ kraken-build --db STANDARD --build --threads 32
```

[return to top](#top)

### > Hmmer {#hmmer}

* <font color="#808080">Supports:</font> [Prokka](#prokka).
* <font color="#00FF80">Simple:</font> Simple compilation and installation.

Download the latest version (v3.1b2) from [this link](http://eddylab.org/software/hmmer3/3.1b2/hmmer-3.1b2-linux-intel-x86_64.tar.gz), or visit its [project page](http://hmmer.org/).

For installation:

```
$ tar -zxvf hmmer-3.1b2-linux-intel-x86_64.tar.gz
$ cd hmmer-3.1b2-linux-intel-x86_64
$ ./configure --prefix="$HOME/local"
$ make
$ make install
```

[return to top](#top)

### > HTSlib {#htslib}

* <font color="#808080">Supports:</font> [Bcftools](bcftools), [SAMtools](#samtools).
* <font color="#00FFFF">Dependencies:</font> [Bzip2](#bzip2), [XZ Utils](#xzutils).

[Bcftools](#bcftools), [HTSlib](#htslib) and [SAMtools](#samtools) come as a suite. If you try to install more than one of them, it is better to install the same version across. The repository page on GitHub is [here](https://github.com/SAMtools). The latest version so far is 1.5, on release page [here](https://github.com/SAMtools/htslib/releases).

For installation:

```
$ tar -jxvf htslib-1.5.tar.bz2
$ cd htslib-1.5
$ ./configure --prefix="$HOME/local" CFLAGS="-I$HOME/local/include" LDFLAGS="-L$HOME/local/lib"
$ make
$ make install
```

[return to top](#top)

### > Infernal {#infernal}

* <font color="#808080">Supports:</font> [Prokka](#prokka).

Project home is [here](http://eddylab.org/infernal/) and you can download the latest relase source code via [this link](http://eddylab.org/infernal/infernal-1.1.2.tar.gz). Make sure you install version >= 1.1 anyway.

For installation:

```
$ tar -zxvf infernal-1.1.2.tar.gz
$ cd infernal-1.1.2
$ ./configure --prefix="$HOME/local"
$ make
$ make install
```

[return to top](#top)

### > Jellyfish {#jellyfish}

* <font color="#808080">Supports:</font> [Kraken](#kraken).
* <font color="#00FF80">Simple:</font> Simple compilation and installation.

Latest release of Jellyfish is 2.2.6, however, Kraken specifically requires Jellyfish v1.x. The latest comparable version (v1.1.11) source codes can be found with [this download](http://www.cbcb.umd.edu/software/jellyfish/jellyfish-1.1.11.tar.gz). Also visit [GitHub](https://github.com/gmarcais/Jellyfish) repository for newer versions.

For installation:

```
$ tar -zxvf jellyfish-1.1.11.tar.gz
$ cd jellyfish-1.1.11
$ ./configure --prefix="$HOME/local"
$ make
$ make install
```

[return to top](#top)

### > GNU Parallel {#parallel}

* <font color="#808080">Supports:</font> [Prokka](#prokka).
* <font color="#00FF80">Simple:</font> Simple compilation and installation.

GNU Parallel is a library belongs to the GNU project. You can find its project home on [GNU website](https://www.gnu.org/software/parallel/), or take a look at its [FTP repository](ftp://ftp.gnu.org/gnu/parallel/), or directly download the [latest release](ftp://ftp.gnu.org/gnu/parallel/parallel-20170722.tar.bz2). Make sure you install version >= 20130422 anyway (latest version appears on bottom of this page).

For installation:

```
$ tar -jxvf parallel-20170722.tar.bz2
$ cd parallel-20170722
$ ./configure --prefix="$HOME/local"
$ make
$ make install
```

[return to top](#top)

### > Prodigal {#prodigal}

* <font color="#808080">Supports:</font> [Prokka](#prokka).
* <font color="#00FF80">Simple:</font> Simple compilation and installation.

Download the source code [here](https://github.com/hyattpd/Prodigal/archive/v2.6.3.tar.gz) via [GitHub](https://github.com/hyattpd/Prodigal/).

For installation:

```
$ tar -zxvf Prodigal-2.6.3.tar.gz
$ cd Prodigal-2.6.3
$ make install INSTALLDIR="$HOME/local/bin"
```

[return to top](#top)

### > Prokka {#prokka}

* <font color="#00FFFF">Dependencies:</font> [Aragorn](#aragorn), [BLAST+](#blast), [Hmmer](#hmmer), [Infernal](#infernal), [GNU Parallel](#parallel), [Prodigal](#prodigal), [tbl2asn](#tbl2asn), [XML::Simple](xml_simple).
* <font color="#00FF80">No-compile:</font> No need to compile.
* <font color="#FF8000">Special:</font> Need to build databases after successful installation.

The latest version in v1.1.2 via [this link](http://www.vicbioinformatics.com/prokka-1.12.tar.gz) found at [project home](http://www.vicbioinformatics.com/software.prokka.shtml).

For installation (simply untar and copy):

```
$ tar -zxvf prokka-1.12.tar.gz
$ mv prokka-1.12/ ~/local/prokka
```

Don't forget to add it to your PATH environment variable, if you don't know how to do it, check below:

```
$ echo 'export PATH="$HOME/local/prokka/bin:$PATH"' >> ~/.bashrc
```

Logout then login again, then build the database with:

```
$ prokka --setupdb
```

[return to top](#top)

### > SAMtools {#samtools}

* <font color="#808080">Supports:</font> [Bcftools](#bcftools), [Pysam](#pysam).
* <font color="#00FFFF">Dependencies:</font> [HTSlib](#htslib).

[Bcftools](#bcftools), [HTSlib](#htslib) and [SAMtools](#samtools) come as a suite. If you try to install more than one of them, it is better to install the same version across. The repository page on GitHub is [here](https://github.com/SAMtools). The latest version so far is 1.5, on release page [here](https://github.com/SAMtools/SAMtools/releases). SAMtools are already built on Discovery cluster, however, it is relatively low in version (0.1.19).

For installation:

```
$ tar -jxvf SAMtools-1.5.tar.bz2
$ cd SAMtools-1.5
$ ./configure --prefix="$HOME/local" CFLAGS="-I$HOME/local/include" LDFLAGS="-L$HOME/local/lib" --with-htslib="$HOME/local" --without-curses
$ make
$ make install
```

[return to top](#top)

### > tbl2asn {#tbl2asn}

* <font color="#808080">Supports:</font> [Prokka](#prokka).
* <font color="#00FF80">No-compile:</font> No need to compile.

The pre-compiled binary file is available on NCBI via [this link](ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools/converters/by_program/tbl2asn/linux64.tbl2asn.gz) or check out [this FTP](ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools/converters/by_program/tbl2asn/)

For installation (simply gunzip, chmod and copy):

```
$ gunzip -d linux64.tbl2asn.gz
$ chmod 755 linux64.tbl2asn
$ mv linux64.tbl2asn ~/local/bin/tbl2asn
```

[return to top](#top)

### > XZ Utils {#xzutils}

* <font color="#808080">Supports:</font> [HTSlib](#htslib).

Visit [project home](https://tukaani.org/xz/) and download latest version (v5.2.3) [here](https://tukaani.org/xz/xz-5.2.3.tar.gz), or clone from [GitHub](git clone https://git.tukaani.org/xz). This library is also knwos as LZMA.

For installation:

```
$ tar -zxvf xz-5.2.3.tar.gz
$ cd xz-5.2.3
$ ./configure --prefix="$HOME/local"
$ make
$ make install
```

[return to top](#top)

### > Zlib {#zlib}

* <font color="#808080">Supports:</font> [Bedtools](#bedtools).
* <font color="#FF0000">Attention:</font> Compiling with special attention.

Visit [project home](https://zlib.net/) and download latest version (v1.2.11) [here](https://zlib.net/zlib-1.2.11.tar.gz). Zlib is actually already built on Discovery cluster, however, bedtools specifically requires a static library (libz.a) to link, which cannot be found on Discovery cluster. You can use my build (/home/li.gua/usr/zlib/1.2.11/lib/libz.a) instead of building your own. Since it is a static library, once bedtools is linked, it is no longer required.

For installation:

```
$ tar -zxvf zlib-1.2.11.tar.gz
$ cd zlib-1.2.11
$ make --64 --prefix="$HOME/local"
$ make install
```

<font color="#FF0000">Attention:</font> The --64 flag should be explicitly assigned.

[return to top](#top)

### > Picard (Java) {#picard}

* <font color="#FF8000">Special:</font> This is a jar archive.

Latest release can be found on [GitHub](https://github.com/broadinstitute/picard/releases/tag/2.10.9) (make sure you download the picard.jar, not source code).

Jar file is ready to use, just put it some where you can find. For example:

```
$ if [[ !-d ~/Jar ]]; then mkdir ~/Jar; fi
$ mv picard.jar ~/Jar/
```

When you want to run the jar file, you should load the jre (Java run time):

```
$ module load oracle_java_1.8u60
$ java -jar ~/Jar/picard.jar -h # this shows the help
```

[return to top](#top)

### > Anvio (Python3) {#anvio}

* <font color="#FF8000">Special:</font> This is a Python module.
* <font color="#FF0000">Attention:</font> This library only works for Python3.

For installation:

```
$ module load python-3.5.2
$ export PYTHONPATH="$HOME/local/lib/python3.5/site-packages"

pip3 install anvio --upgrade --prefix="$HOME/local"
```

For use we must load Python3 (recommend <b>DO NOT</b> load Python module with .bashrc file):

```
$ module load python-3.5.2
$ export PYTHONPATH="$HOME/local/lib/python3.5/site-packages"

# put your script here ###
```

[return to top](#top)

### > CONCOCT (Python2) {#concoct}

* <font color="#FF8000">Special:</font> This is a Python module.
* <font color="#00FFFF">Dependencies:</font> [Pysam](#pysam).
* <font color="#FF0000">Attention:</font> This library only works for Python2.
* <font color="#FF0000">Attention:</font> Special components should be installed.

Due to the pip bug (specifically in this version on Discovery cluster, not anyone), we have to install the module manually from source code. Download from [GitHub](https://github.com/BinPro/CONCOCT/archive/0.4.0.tar.gz).

```
$ module load python-2.7.5
$ export PYTHONPATH="$HOME/local/lib/python2.7/site-packages"

$ tar -zxvf CONCOCT-0.4.0.tar.gz
$ cd CONCOCT-0.4.0
$ python setup.py install --prefix="$HOME/local"

$ cp scripts/ ~/Programs/CONCOCT/ # copy all scripts
$ cp scgs/ ~/Programs/CONCOCT/scgs/
```

The Cdd to Cog table is out-of-date with the newer versions of Cdd. It should be updated with a little perl:

```
# the mapping table from Cdd to Cog:
$ cd ~/Programs/CONCOCT/scgs/
$ wget ftp://ftp.ncbi.nih.gov/pub/mmdb/cdd/cddid_all.tbl.gz
$ gunzip cddid_all.tbl.gz
$ cut -f 1-2 cddid_all.tbl | perl -ne 'chomp $_;if($_=~/COG\d+$/){print($_,"\n");}' > cdd_to_cog.tsv
```

This installation will put a exceutable script into ~/local/bin, (no need to do 'import concoct' within python). For calling concoct:

```
$ module load python-2.7.5
$ export PYTHONPATH="$HOME/local/lib/python2.7/site-packages"

$ concoct --help
```

[return to top](#top)

### > Pysam (Python) {#pysam}

* <font color="#FF8000">Special:</font> This is a Python module.
* <font color="#808080">Supports:</font> [CONCOCT](#concoct).
* <font color="#00FFFF">Dependencies:</font> [SAMtools](#samtools).

Source code packages can be found on [GitHub](https://github.com/SAMtools/htslib).

<font color="#00FF00"><b>Note:</b></font> This module is intended to support CONCOCT, thus I installed it with python2.7:

```
$ module load python-2.7.5
$ export PYTHONPATH="$HOME/local/lib/python2.7/site-packages"

$ git clone https://github.com/pysam-developers/pysam.git
$ cd pysam
$ python setup.py build
$ python setup.py install --prefix="$HOME/local"
```

[return to top](#top)

### > Quast (Python) {#quast}

* <font color="#FF8000">Special:</font> This is a Python module.

<font color="#00FF00"><b>Note:</b></font> This module is inteded to work with both Python2.7+ and Python3.3+. I installed it into Python3, you can choose what you prefer. Here is my installation example using pip:

```
$ module load python-3.5.2
$ export PYTHONPATH="$HOME/local/lib/python3.5/site-packages"

pip3 install quast --upgrade --prefix="$HOME/local"
```

This installation will put a exceutable script into ~/local/bin, (no need to do 'import quast' within python). For calling quast:

```
$ module load python-3.5.2
$ export PYTHONPATH="$HOME/local/lib/python3.5/site-packages"

$ quast --help
```

[return to top](#top)

### > List::MoreUtils (Perl) {#list_moreutils}

* <font color="#808080">Supports:</font> [BLAST+](#blast).
* <font color="#FF8000">Special:</font> This is a perl module. Make sure you loaded module perl-5.20.0 before procceding.

For installation:

First, open cpan (answer 'yes' when asked about automation) and input commands below (this only needs to be done once). These commands will assign a local location for building perl modules:

```
$ cpan

cpan> o conf mbuildpl_arg "--install_base /home/your_account_name/local/"
cpan> o conf makepl_arg "PREFIX=/home/your_account_name/local/"
cpan> o conf commit
```

Then in cpan, install this modue by:

```
cpan> force install List::MoreUtils
```

Also do not forget to export the local library path by this shell command (this only needs to be done once):

```
$ echo "export PERL5LIB=/home/your_account_name/local/share/perl5:/home/your_account_name/local/site_perl/5.20.0:$PERL5LIB" >> ~/.bashrc
```

[return to top](#top)

### > XML::Simple (Perl) {#xml_simple}

* <font color="#808080">Supports:</font> [Prokka](#prokka).
* <font color="#FF8000">Special:</font> This is a perl module. Make sure you loaded module perl-5.20.0 before procceding.

For installation:

First, open cpan (answer 'yes' when asked about automation) and input commands below (this only needs to be done once). These commands will assign a local location for building perl modules:

```
$ cpan

cpan> o conf mbuildpl_arg "--install_base /home/your_account_name/local/"
cpan> o conf makepl_arg "PREFIX=/home/your_account_name/local/"
cpan> o conf commit
```

Then in cpan, install this modue by:

```
cpan> force install XML::Simple
```

Also do not forget to export the local library path by this shell command (this only needs to be done once):

```
$ echo "export PERL5LIB=/home/your_account_name/local/share/perl5:/home/your_account_name/local/site_perl/5.20.0:$PERL5LIB" >> ~/.bashrc
```

[return to top](#top)

## Utility installation guide

For those people need to use newer versions of GCC, as (assembler) and ld (linker) for compilation, I provided the installation I did below:

### GCC 5.4.0 (Optional) {#gcc}

Download GCC suite from [GNU FTP](ftp://ftp.gnu.org/gnu/gcc/) (I downloaded version 5.4.0, do not build too new versions, some times they too, cause problems). Then:

```
$ tar -zxvf gcc-5.4.0.tar.gz
$ cd gcc-5.4.0
$ contrib/download_prerequisites # this downloads the prerequisites like gmp, mpfr, etc., and automatically build with gcc.
$ mkdir build_dir && cd build_dir
$ ../configure --prefix="$HOME/usr/gcc/5.4.0" --enable-languages=c,c++,fortran --disable-multilib # you can also add ada and go if needed
$ make -j 32 # allocate a 32-CPU node to do this will be a lot faster
$ make check -j 32 # I encountered a failure, in zmp or something, but not critical, so ignored
$ make install
```

<font color="#00FF00"><b>Note:</b></font> Make sure you override the system's default bin path with PATH env-var.

[return to top](#top)

### Binutils (Optional) {#binutils}

Binutils are essential to compilers but not released in GCC suite. The two most imporatant ones are as (assembler) and ld (linker). The as and ld version on Discovery cluster is extremely out-of-date (released in 2010; compare to GCC 5.2.1, in 2015). This will cause some problems in compiling newer software. Download GNU binutils from [GNU FTP](ftp://ftp.gnu.org/gnu/binutils/) (I downloaded version 2.26.1). Then:

```
$ tar -zxvf binutils-2.26.1.tar.gz
$ cd binutils-2.26.1
$ mkdir build_dir && cd build_dir
$ ../configure --prefix="$HOME/usr/binutils/2.26.1"
$ make -j 32 # do this on interactive node
$ make check -j 32 # always check or test if available
$ make install
```

<font color="#00FF00"><b>Note:</b></font> Make sure you override the system's default bin path with PATH env-var.

[return to top](#top)
